zip-0302.rst

```diff
@@ -3,8 +3,109 @@
   ZIP: 302
   Title: Standardized Memo Field Format
   Owners: Jack Grigg <jack@electriccoin.co>
-  Original-Authors: Jay Graber
-  Status: Reserved
+  Original-Author: Jay Graber
+  Status: Draft
   Category: Standards / RPC / Wallet
+  Created: 2017-02-08
   Discussions-To: <https://github.com/zcash/zips/issues/366>
   Pull-Request: <https://github.com/zcash/zips/pull/105>
+
+Abstract
+========
+
+This ZIP describes a proposed specification for a standardized format for clients who wish
+to transmit or receive content within the encrypted memo field of shielded transactions.
+
+Motivation
+==========
+
+A well-defined standard for formatting content within the encrypted memo field will help
+expand its use within the Zcash ecosystem by providing a commonly recognized format for  
+memo values carrying different types of data. Users and third-party services benefit from 
+standardized formatting rules that define the type and length of the data contained within.
+
+Specification
+===============
+
+Section 5.5 of the Zcash protocol specification [#protocol]_ defines three cases
+for the encoding of a memo field:
+
+* a UTF-8 human-readable string [#UTF-8]_, padded by appending zero bytes; or
+* the byte ``0xF6`` followed by 511 ``0x00`` bytes, indicating "no memo"; or
+* any other sequence of 512 bytes starting with a byte value ``0xF5`` or greater (which is
+  therefore not a valid UTF-8 string), as specified in ZIP 302.
+
+This ZIP refines the specification of the third case.
+
+The following specification constrains a party, called the "reader", that interprets the
+contents of a memo. It does not define consensus requirements.
+
++ If the first byte (byte 0) has a value of 0xF4 or smaller, then the reader MUST:
+     + strip any trailing zero bytes
+     + decode it as a UTF-8 string (if decoding fails then report an error).
+
++ If the first byte has a value of 0xF5, then the reader MUST:
+     + Interpret the next few bytes (1 to 9 of them) as a 64-bit unsigned variable-length
+       integer [#Bitcoin-CompactSize]_, and use it as an arbitrary application-defined
+       "type" field.
+     + Interpret the next 1 or 2 bytes as a CompactSize value constrained to the range 0..510, and use 
+       the resulting value as the length of the remaining data.
+     + If 1 + the number bytes used for the type field + the number of bytes used for the
+       length field + the length > 512 then report an error.
+     + Inspect the padding after the end of the indicated length, and if it
+       contains anything other than bytes of value 0x00 then report an error.
+     + Return to the caller a 3-tuple of the following data:
+           + the type — an integer in :math:`[0...2^{64})`
+           + the length — an integer in :math:`[0...510)`
+           + a byte string of that length that contains the payload
```

Comment:
We should document prior to merging the ZIP draft that this is not agreed upon yet, due to incomplete discussion in https://github.com/zcash/zcash/issues/1849, so implementations should not assume any standard related to prefix `0xF5`.

Reply:
Should we create a separate PR that adds this section, and remove it from this PR?

Reply:
Yeah, I prefer that; I'll make the change.

Reply:
```suggestion
```

Reply:
I have now created the separate PR: https://github.com/zcash/zips/pull/638